docs/src/developers/style.md

```diff
@@ -51,6 +51,220 @@ format("test")
+```julia
+_internal_function(x::Integer) = x + 1
+function _internal_function(x)
+    error(
+        "Internal error. This probably means that you called " *
+        "public_function() with the wrong type.",
+    )
+end
+public_function(x) = _internal_function(x)
+```
+
+#### Dealing with correctness
+
+Dealing with correctness is harder, because Julia has no way of formally
+specifying interfaces that abstract types must implement. Instead, here are two
+options that you can use when writing and interacting with generic code:
+
+**Option 1: Use concrete types and let users extend new methods.**
+
+In this option, explictly restrict input arguments to concrete types that
+are tested and have been validated for correctness. For example:
+```jldoctest my_sum
+julia> function my_sum_option_1(x::Vector{Float64})
+           y = 0.0
+           for i in 1:length(x)
+               y += x[i]
+           end
+           return y
+       end
+my_sum_option_1 (generic function with 1 method)
+
+julia> my_sum_option_1([1.0, 2.0, 3.0])
+6.0
+```
+
+Using concrete types satisfies the `MethodError` principle:
+```jldoctest my_sum
+julia> my_sum_option_1("abc")
+ERROR: MethodError: no method matching my_sum_option_1(::String)
+```
+and it allows other types to be supported in future by defining new methods:
+```jldoctest my_sum
+julia> function my_sum_option_1(x::Array{T,N}) where {T<:Number,N}
+           y = zero(T)
+           for i in eachindex(x)
+               y += x[i]
+           end
+           return y
+       end
+my_sum_option_1 (generic function with 2 methods)
+```
+Importantly, these methods do not have to be defined in the original package.
+
+!!! info
+    Some usage of abstract types is okay. For example, in `my_sum_option_1`, we
+    allowed the element type, `T`, to be a subtype of `Number`. This is fairly
+    safe, but it still has an implicit assumption that `T` supports `zero(T)`
+    and `+(::T, ::T)`.
+
+**Option 2: Program defensively, and validate all assumptions.**
+
+An alternative is to program defensively, and to rigorously document and
+validate all assumptions that the code makes. In particular:
+
+ 1. All assumptions on abstract types that aren't guaranteed by the definition
+    of the abstract type (for example, optional methods without a fallback)
+    should be documented.
+ 2. If practical, the assumptions should be checked in code, and informative
+    error messages should be provided to the user if the assumptions are not
+    met.
+ 3. Tests should cover for a range of corner cases and argument types.
+
+For example:
+```jldoctest my_sum
+"""
+    my_sum_defensive(x::AbstractArray{T}) where {T}
+
+Return the sum of the elements in the abstract array `x`.
+
+## Assumptions
+
+This function makes the following assumptions:
+
+ * That `zero(T)` is defined
+ * That `x` supports the iteration interface
+ * That  `+(::T, ::T)` is defined
+"""
+function my_sum_defensive(x::AbstractArray{T}) where {T}
+    try
```

Comment:
We don't want to start doing this in JuMP and MOI right ? Documenting it is enough I think

Reply:
No. At the JuMP level, we mostly go with the concrete type suggestion.

But here's a case where we (I) didn't check assumptions: https://github.com/jump-dev/JuMP.jl/issues/2999